### ISSUE TITLE
Value of "Storage Charge/Discharge Mode"

### ISSUE THREAD

**Issue by CaptainVoni** [biemond/solaredge.modbus#47]
The value of "Storage Charge/Discharge Mode" is defined as text, so I assumed it will return the label of the respective mode. But instead it returns a number, e.g. in case of Maximize self-consumption it will contain "7" as value.

Is this intentionally? If so, how can we verify if a specific mode is set, i.e. is there somewhere a documented mapping between the label and the number?

Shouldnt it be possible to set the mode to a specific value and when verifying the mode, that same value is returned?

**Comment by biemond** [comment]
Hey, 

you can check here. for me it is a enum. don't know why homey does not show this https://github.com/biemond/solaredge.modbus/blob/main/.homeycompose/capabilities/storagedefaultmode.json

and it is a number on solaredge not a text.

**Comment by biemond** [comment]
I think they do this because of locale and when you switch language it should still work and don't get in trouble.

**Comment by CaptainVoni** [comment]
If I understand you correct it is not an issue of this app, but of Homey. I guess I should report it there then, correct?

I understand your point regarding switching the language, but it should be possible to solve this in some other way. It just doesnt make any sense, when you set some value to X, then retrieve that value and you get Y. That's just begging for troubles in my opinion.

**Comment by biemond** [comment]
I think you used some logical value or not. this is direct on the capability.

I can add it , what do you need?  some  "And condition" so you can compare.  you need to provide the state and give back true / false.   Or you can get status changed trigger but that is only possible when you trigger it.

**Comment by CaptainVoni** [comment]
That's exactly what I need. In some flows I need to know the current value of "Storage Charge/Discharge Mode". I could use the logic flow card "<Text/number> is equal to <value>", but the logical value is a number, while if I use the Then-flow-card "Storage Charge/Discharge Mode" to set the mode, it expects a text.

The same is (at least) also true for "Storage Control Mode": you set it as text, but when you retrieve it, you will get a number.

And btw: maybe adding a "Set " at the beginning of the Then-flow-card for e.g.  "Storage Charge/Discharge Mode" and " to " at the end would make the flows more readable. Example: "Set Storage Charge/Discharge Mode to" instead of just "Storage Charge/Discharge Mode" for the flow card. This way it would also be easier to distinguish between the flow cards and the values/capabilities - right now they have the same name.

**Comment by CaptainVoni** [comment]
And: who is providing the translations? Right now German is a big mess - some English, some German - I could provide proper translations for the cards/variables if you want.

**Comment by biemond** [comment]
interesting I don't do german , only english and some dutch. So I guess for the default stuff , homey try to do the locale.

**Comment by biemond** [comment]
this is the main 1 but it generated from other places https://github.com/biemond/solaredge.modbus/blob/main/app.json but at least you get the idea

**Comment by CaptainVoni** [comment]
Never used github, so I'm not sure how's the best way to provide you the translations. Should I just edit the file ("Edit in place") and you accept/reject the changes?

**Comment by biemond** [comment]
you can download it and upload it here.  but it will be a lot of work

**Comment by CaptainVoni** [comment]
See the attached file (as zip as json is not accepted here). I translated just the triggers for now, because before I translate everything I want to make sure that it's ok for you how I'm doing it. I may update some of this translations when I do the rest simply for consistency.

[app.zip](https://github.com/biemond/solaredge.modbus/files/12346561/app.zip)

**Comment by CaptainVoni** [comment]
Should I interpet your thumbs up as "Do it"? ;-)

**Comment by biemond** [comment]
will check it in the weekend and will make a test version for you which you can verify.

**Comment by biemond** [comment]
here we go https://homey.app/a/solaredge.modbus/test/ and thanks again.

**Comment by CaptainVoni** [comment]
Worked like a charm. The translations are - as expected - pretty inconsistent, so I have to take some time to sort that out.

Is it correct, that this test version only contains the german translations, but not the "and condition" we discussed? If it is included, I can't find it.

Where did you get the labels for the charge/discharge modes for the Solaredge inverters, e.g. "Charge excess PV power only"? They are different from the labels Solaredge itself uses in their [documentation](https://knowledge-center.solaredge.com/sites/kc/files/storedge_charge_discharge_profile_programming.pdf). Should I update them (the english ones) to fit the labels from Solaredge? (Especially the "Charge excess PV power only" is pretty misleading in my opinion)

And one final question if I may: where did the values that may be selected for "Set Discharge power" come from? I'm asking because I would need 0 as value (or 1 as fallback). And before you ask: no, there is no fitting battery mode for my case ;-)

I will start working on the translations in the coming days and let you know as soon as I'm finished.

**Comment by biemond** [comment]
condition is still to do. 

everything comes from their modbus doc.

<img width="1203" alt="Screen Shot 2023-08-20 at 14 06 11" src="https://github.com/biemond/solaredge.modbus/assets/1085004/4290e34d-e56c-4cca-937a-4c963ed8e223">
<img width="1178" alt="Screen Shot 2023-08-20 at 14 06 24" src="https://github.com/biemond/solaredge.modbus/assets/1085004/e0a4b244-21f5-4b97-b801-062062b64dbe">

**Comment by CaptainVoni** [comment]
Interesting... they use different labels in different documents for the same functionality. Do you have a link to that document?

**Comment by biemond** [comment]
I calculate those vaules because it goes over 2 register .  so made it easier for myself  and easy to calculate those values

https://github.com/biemond/solaredge.modbus/blob/main/drivers/invertorwithbatt/driver.flow.compose.json#L629

**Comment by biemond** [comment]
[Power Control Open Protocol for SolarEdge Inverters.pdf](https://github.com/biemond/solaredge.modbus/files/12389010/Power.Control.Open.Protocol.for.SolarEdge.Inverters.pdf)

**Comment by biemond** [comment]
profiles I cannot support and it is done by solaredge NL. I know on BE some can do this. plus I don't have installation account. I know here they can do this and you need to use a fake business name https://www.solar-bouwmarkt.nl/solaredge-omvormers-solaredge-monitoring-portal-in.html and solaredge is not happy with this.

**Comment by CaptainVoni** [comment]
> I calculate those vaules because it goes over 2 register . so made it easier for myself and easy to calculate those values
So it's not possible to set the discharging power to zero or do you have to make some calculations?

**Comment by biemond** [comment]
yes 0 is possible. was thinking I already supported that. will add that

**Comment by biemond** [comment]
oh it was for import / export limit https://github.com/biemond/solaredge.modbus/blob/main/drivers/invertorwithbatt/device.ts#L299

**Comment by CaptainVoni** [comment]
> profiles I cannot support and it is done by solaredge NL. I know on BE some can do this. plus I don't have installation account. I know here they can do this and you need to use a fake business name https://www.solar-bouwmarkt.nl/solaredge-omvormers-solaredge-monitoring-portal-in.html and solaredge is not happy with this.

Didn't really understand this.

**Comment by biemond** [comment]
like this
[storedge_charge_discharge_profile_programming.pdf](https://github.com/biemond/solaredge.modbus/files/12389089/storedge_charge_discharge_profile_programming.pdf)

**Comment by CaptainVoni** [comment]
I just saw that it was automatically installed last night. Will test it and report back.

FYI: I'm about 3/4 through the translations; I guess I will finish in the next days.

**Comment by CaptainVoni** [comment]
Here my results:
- Setting the discharge power to zero results in resetting it to 5000W (seems to be the default), i.e. no effect.
- Setting the charge power to zero works fine, i.e. charging is prevented.
- Querying the charge/discharge mode works fine.

**Comment by biemond** [comment]
thanks , shall  we do 50w discharge instead?

**Comment by CaptainVoni** [comment]
May I vote for one Watt? We should have the absolute minimum as option. 

In case you want to know why: if the price from the grid is lower than from the battery, the inverter should create AC for self consumption, then charge the battery and feed any excess to the grid, but if the PV yield is too low to cover the consumption it should **not** use the battery. There is no charge/discharge mode for this scenario. Right now I switch between Max Self Consumption and Off depending on the PV yield, but to do so I need to trigger a flow whenever consumption or PV yield changes (i.e. a LOT!). If I could set the max discharge power to a very low value, I could use simply Max Self Consumption with the discharge power set.

**Comment by biemond** [comment]
yep lets try that.

**Comment by biemond** [comment]
in that case I would put active power to 0 and import it all from grid and don't do own consumption.

**Comment by CaptainVoni** [comment]
> in that case I would put active power to 0 and import it all from grid and don't do own consumption.

But if you set Active Power to 0, you would lose the complete PV yield, as it just shuts down the PV inverter:

> Active Power Limit dynamic control: This register controls the active power limit of the inverter dynamically. It is set as the 
percentage of the inverter’s maximum power. For example: when setting the register of SE5000 to 20, it will limit the inverter to 
1000W which is 20% of 5000W.

And I always thought that there are two inverters: one to convert PV-DC to AC and one to convert Battery-DC to AC. This would explain why there are two individual limits for these two. If this is true, setting Active Power to zero wouldnt affect battery discharging.

**Comment by biemond** [comment]
try it out because if the price goes very negative ,  import it all from grid.  else do self consumption and no export. if the price is ok, do self consumption and full export. 

the challenge is winter , to check with solcast or the price , the plan for the next day when to charge the batt and when not.

**Comment by CaptainVoni** [comment]
Just tested it and this obviously is wrong:

> And I always thought that there are two inverters: one to convert PV-DC to AC and one to convert Battery-DC to AC. This would explain why there are two individual limits for these two. If this is true, setting Active Power to zero wouldnt affect battery discharging.

If Active Power Limit is set to zero (and Charge/Discharge Mode is set to Maximize Self Consumption and the battery is fully charged), all the power is taken from the grid. So Active Power Limit somehow limits also battery discharging, but it does not limit charging. I guess, there is only one DC-AC inverter, which is used for battery and PV together.

> in that case I would put active power to 0 and import it all from grid and don't do own consumption.

I guess you assume that the price for electricity **from** the grid is identical to the price **to** the grid at my place, which isn't the case. The price from the grid changes hourly and is negative from time to time; the price to the grid is determined monthly and has never been negative or even close to zero.
Also the price for electricity out of the battery is quite high if you consider the cost for the battery. Right now I calculate 0.217€/kWh for electricity out of the battery.
So whenever the grid price is below 0.217€, I want the battery to be charged (if possible), but not discharged (grid is cheaper). At the same time the PV production should not be restricted, because feeding excess PV production to the grid earns me money.

There might be some situations when it would be advantageous to disable PV production with Active Power Limit, but not only does the grid price has to be (very) negative, the consumption also has to be high. I can think of some rules for that, but they are quite complex. Thank you for giving me something to think about the next weeks :-D

**Comment by CaptainVoni** [comment]
> the challenge is winter , to check with solcast or the price , the plan for the next day when to charge the batt and when not.

That's on my TODO list. Estimate how long the battery will last (done) and then charge it from the grid at the cheapest hour within that period.

**Comment by biemond** [comment]
yep , here in NL it is around -19c , we had sometimes this summer -49c . yes it is unbelievable how much tax is added. I think we have added tax + energy tax and transport cost.

**Comment by CaptainVoni** [comment]
I'm finished with the translations, but I'm uncertain with regard to "Total Yield" and "Total Day Yield": to me it seems that this isn't a "yield" but more a "consumption". E.g. meter_power.daily is displayed in Homey as Total Yield, but it changes when there is no PV production. Also judged purely from "meter_power", this should be the energy measured by the house meter, i.e. the electricity drawn from the grid.

I'm definitively no english native speaker, but at least to me this is confusing.

Do you have an opinion on this? Is the term "yield" correct in your view?

**Comment by biemond** [comment]
I think today yield and day yield is correct.  don't see how  meter_power.daily is displayed in Homey as Total Yield. it show correct for me 
no it is total yield not consumption or smart meter. you have losses on everything etc.  but it is what measured and what should be possible. 
What does yield mean in solar? Energy yield is the amount of energy actually harvested from solar panels, taking into consideration external factors like heat, dirt, and shade, whereas efficiency refers to testing done in lab conditions.

<img width="375" alt="Screen Shot 2023-08-28 at 00 07 43" src="https://github.com/biemond/solaredge.modbus/assets/1085004/6c1cfb6b-7798-4f3f-b463-89df9d4fb04c">
<img width="389" alt="Screen Shot 2023-08-28 at 00 10 02" src="https://github.com/biemond/solaredge.modbus/assets/1085004/dc8503ca-7ca0-4a1b-b081-554e2b400f2c">

**Comment by CaptainVoni** [comment]
I'm a bit confused about that total day yield, because it shows (very) different numbers from what I see in the Solaredge app. But lets stick with "Yield"/"Ertrag".

See the attached [app.zip](https://github.com/biemond/solaredge.modbus/files/12451992/app.zip) for the complete german translation. As discussed, the file includes five lines starting with "TODO:" with some remarks/questions from me.

Let me know if anything is unclear.

**Comment by biemond** [comment]
yep solaredge calculates it or has more data which we don't have this.

**Comment by biemond** [comment]
done , can you test again https://homey.app/a/solaredge.modbus/test/

**Comment by CaptainVoni** [comment]
Zis looks all weri German too me! 😄
Translations looks great, a handful flow cards are missing translation or the translation is missing „Setze“ at the beginning. I’ll download the latest json and try to figure out what went wrong.

Setting the discharge power to 10 W again doesn’t work - it is set to the 5000 W. 😞 Is there nothing in the documentation which specifies the accepted minimum?

Thank you for your effort!

**Comment by CaptainVoni** [comment]
Yeah, I already know that you updated it. ;-)
Suddenly all the references in flows to the inverter were gone. After a reload they were back. :-)

Testing now.

**Comment by biemond** [comment]
here we go , new version https://homey.app/a/solaredge.modbus/test/

**Comment by biemond** [comment]
I added voltage and now this version has your locale

**Comment by CaptainVoni** [comment]
Checked the voltages and verified all titles (English and German):
* All the voltages "measure_voltage.phaseX" are reported as value around 400V, which doesn't make much sense for a single-phase value. In contrast, the voltages "measure_voltage.meter_phaseX" report values around 230V, which seems to be fine.
* One capability had a title for an action. Fixed that.
* I added recommendations for the German and English titles of all actions in order to be in-line with the general titles of actions in Homey.

[app-3.0.43 (updated).zip](https://github.com/biemond/solaredge.modbus/files/12576485/app-3.0.43.updated.zip)

**Comment by biemond** [comment]
yes meter ones are beter , for me somehow it is 240 for phase 1. 

All the voltages "measure_voltage.phaseX" are reported as value around 400V, which doesn't make much sense for a single-phase value

**Comment by biemond** [comment]
pushed a new version

**Comment by biemond** [comment]
thanks , yep can you find out what minimal discharge works for you , from the inverter menu.

**Comment by CaptainVoni** [comment]
> thanks , yep can you find out what minimal discharge works for you , from the inverter menu.

What do you mean by "inverter menu"? I just checked everything I can find in SetApp and also on the monitoring website, but I can't find anything about the discharge power of the battery.

**Comment by biemond** [comment]
ok will do some checks

**Comment by CaptainVoni** [comment]
I guess I should report this to the Homey support, right?

![image](https://github.com/biemond/solaredge.modbus/assets/1625329/f7c77ca3-ce53-4ee1-beca-cc30172bea5c)

- When I use a logic card to check if the battery discharge power ("Batterieentladeleistung") is greater than zero, it is not (correct).
- When I use the App flow card to check if the battery discharge power is greater than zero, it is (incorrect).
- The battery discharge power actually is zero. 

I'm under the impression that the Homey built-in flow cards for capabilities are buggy.

**Comment by biemond** [comment]
oh let me verify this.  thanks for reporting.

**Comment by biemond** [comment]
indeed I see the same, always true

**Comment by biemond** [comment]
So greater than works for me, but you cannot test them from advanced flow. there is for me , no true or false output. and then the value is not tested

<img width="1047" alt="Screen Shot 2023-09-23 at 18 32 18" src="https://github.com/biemond/solaredge.modbus/assets/1085004/4f81978c-2524-4604-9320-8b32ea1edeae">

**Comment by biemond** [comment]
so discharge 0 ,10,50 and 100 is working for me

<img width="498" alt="Screen Shot 2023-09-23 at 20 00 56" src="https://github.com/biemond/solaredge.modbus/assets/1085004/529c0816-5f60-478b-b155-e17c67ad16
<img width="436" alt="Screen Shot 2023-09-23 at 19 50 32" src="https://github.com/biemond/solaredge.modbus/assets/1085004/4d88a0b5-5cfa-4da3-a73c-72ed216b1d20">
<img width="410" alt="Screen Shot 2023-09-23 at 19 50 38" src="https://github.com/biemond/solaredge.modbus/assets/1085004/efacf31b-a09d-4880-b9b8-b8449558628f">
<img width="405" alt="Screen Shot 2023-09-23 at 19 56 20" src="https://github.com/biemond/solaredge.modbus/assets/1085004/ea3628cd-23cc-48d1-812e-7ed6174a8207">
<img width="391" alt="Screen Shot 2023-09-23 at 19 58 53" src="https://github.com/biemond/solaredge.modbus/assets/1085004/ce7338be-51ff-4920-be03-0222a68fda70">
<img width="356" alt="Screen Shot 2023-09-23 at 20 00 49" src="https://github.com/biemond/solaredge.modbus/assets/1085004/87a8fc93-5517-4a9c-8038-d14fadc27c8e">
0c">

**Comment by CaptainVoni** [comment]
> so discharge 0 ,10,50 and 100 is working for me

Tested all of them right now: all are working for me except zero. But 10 W is also perfectly fine for me. Many thanks for your efforts! When do you think will you release this version?

**Comment by CaptainVoni** [comment]
Are you planning on adding such trigger?

>     * A trigger for a change of the storage control mode and the storage charge/discharge mode is missing. If you want to add them you can use "Speicherkontrollmodus hat sich geändert" (storage control mode) and "Speicher-Lade/Entlade-Modus hat sich geändert" (storage charge/discharge mode) as translation.

I'm asking because my battery turns itself off every day between 21:00 and 22:00 and I want to react on that.

**Comment by biemond** [comment]
no problem. also I think the default of remote action = 6 hours , than you need to do something else it will go to the default again.

let me ask to publish it to release

**Comment by CaptainVoni** [comment]
> no problem. also I think the default of remote action = 6 hours , than you need to do something else it will go to the default again.

Oh, that's interesting! Will check that. Thanks for the hint!

**Comment by biemond** [comment]
I think it is an hour but I set it to 6 hours

```
  if (type == 'storagedefaultmode') {
        const storagedefaultmodeRes = await client.writeSingleRegister(0xe004, Number(4));
        console.log('controlmodewrite', storagedefaultmodeRes)

        // set timeout to 6 hours, done in seconds
        const remoteTimeout = await client.writeMultipleRegisters(0xe00b, [ Number(21600), 0 ]);
        console.log('remote_control_command_timeout', remoteTimeout);
        // 0 – Off
        // 1 – Charge excess PV power only.
        // Only PV excess power not going to AC is used for charging the battery. Inverter NominalActivePowerLimit (or the inverter rated power whichever is lower) sets how much power the inverter is producing to the AC. In this mode, the battery cannot be discharged. If the PV power is lower than NominalActivePowerLimit the AC production will be equal to the PV power.
        // 2 – Charge from PV first, before producing power to the AC.
        // The Battery charge has higher priority than AC production. First charge the battery then produce AC.
        // If StorageRemoteCtrl_ChargeLimit is lower than PV excess power goes to AC according to NominalActivePowerLimit. If NominalActivePowerLimit is reached and battery StorageRemoteCtrl_ChargeLimit is reached, PV power is curtailed.
        // 3 – Charge from PV+AC according to the max battery power.
        // Charge from both PV and AC with priority on PV power.
        // If PV production is lower than StorageRemoteCtrl_ChargeLimit, the battery will be charged from AC up to NominalActivePow-erLimit. In this case AC power = StorageRemoteCtrl_ChargeLimit- PVpower.
        // If PV power is larger than StorageRemoteCtrl_ChargeLimit the excess PV power will be directed to the AC up to the Nominal-ActivePowerLimit beyond which the PV is curtailed.
        // 4 – Maximize export – discharge battery to meet max inverter AC limit.
        // AC power is maintained to NominalActivePowerLimit, using PV power and/or battery power. If the PV power is not sufficient, battery power is used to complement AC power up to StorageRemoteCtrl_DishargeLimit. In this mode, charging excess power will occur if there is more PV than the AC limit.
        // 5 – Discharge to meet loads consumption. Discharging to the grid is not allowed. 
        // 7 – Maximize self-consumption
        const remotecontrolwrite = await client.writeSingleRegister(0xe00d, Number(value));
        console.log('remotecontrolwrite', remotecontrolwrite);
      }
```

**Comment by CaptainVoni** [comment]
It really might be that timeout. But I do set the mode at least once an hour, but the value to which I set it to might stay the same for some hours.

As I write this, the mode reset itself to off. I checked and I had it set to "Discharge to meet loads consumption" before 18:00 and set it to "Maximize self-consumption" at 18:00. So that's just three hours.

And it resets itself pretty much every day at around the same time. That's really strange...

**Comment by CaptainVoni** [comment]
>         // 5 – Discharge to meet loads consumption. Discharging to the grid is not allowed. 
>         // 7 – Maximize self-consumption

Ok... I need to know... what's mode 6? :-D

**Comment by biemond** [comment]
can you try this , the new change mode triggers https://homey.app/a/solaredge.modbus/test/

**Comment by CaptainVoni** [comment]
Thanks! I just installed it and will keep it testing over the next days. I'll report back.

**Comment by CaptainVoni** [comment]
Just for your info: I'm still testing. Something seems to be fishy, it clearly missed a few state changes yesterday, however my inverter needed a restart yesterday, so I don't want to point fingers in any direction yet.

**Comment by CaptainVoni** [comment]
Correction to my last message: it gets all state changes, but it triggers before the state is updated in the app, i.e. if you read the state directly after the trigger fired, you will get the state BEFORE the state change, not the current one. Verifying this right now and checking how long it takes until the state is updated after the trigger fired.

**Comment by biemond** [comment]
interesting, had to do 2 part , 1 is from the device menu which set and triggers , other is from set Cap and comparing previous value , then fire and set new value . 

maybe I need to fire after the set change is done. 

can you add a 1 sec delay, like to know if 1 sec is enough or set change can take many seconds.

```
if (this.validResultRecord(result['storage_control_mode'])) { 
                    this.addCapability('storagecontrolmode');
                    var storagecontrolmode = result['storage_control_mode'].value;
                    console.log('changedStoragecontrolmode old ' + this.getCapabilityValue('storagecontrolmode') );
                    console.log('changedStoragecontrolmode new ' + storagecontrolmode );
                    if (this.getCapabilityValue('storagecontrolmode') != storagecontrolmode) {
                        let tokens = {
                            "mode": Number(storagecontrolmode)
                        };
                        let state = {};
                        console.log('trigger changedStoragecontrolmode ' + storagecontrolmode );
                        this.homey.flow.getDeviceTriggerCard('changedStoragecontrolmode').trigger(this, tokens, state);
                    }                    
                    this.setCapabilityValue('storagecontrolmode', storagecontrolmode);
                }

                if (this.validResultRecord(result['remote_control_command_mode'])) { 
                    this.addCapability('storagedefaultmode');
                    var storagedefaultmode = result['remote_control_command_mode'].value;
                    if (this.getCapabilityValue('storagedefaultmode') != storagedefaultmode) {
                        let tokens = {
                            "mode": Number(storagedefaultmode)
                        };
                        let state = {};
                        console.log('trigger changedStoragedefaultmode ' + storagedefaultmode );
                        this.homey.flow.getDeviceTriggerCard('changedStoragedefaultmode').trigger(this, tokens, state);
                    } 
                    this.setCapabilityValue('storagedefaultmode', storagedefaultmode);
                }
```

**Comment by CaptainVoni** [comment]
Just a delay of one second is sufficient to reliable get the correct value.

Looking at your code I wonder if the "this.setCapabilityValue(...)" shouldn't be inside the if-clause, just right above the "this.homey.flow.getDeviceTriggerCard(...)...".

**Comment by biemond** [comment]
yes you are correct, let me fix that.  if you do from menu it works

**Comment by CaptainVoni** [comment]
I have the test-version of the app on auto update and have a flow logging all mode-changes to the timeline, so I should see if that really fixes it. I will let you know.

**Comment by biemond** [comment]
yep will push in a hour or so. was working on growatt tl device.

**Comment by CaptainVoni** [comment]
You're pretty busy :smile: I really appreciate the work you put into this - thanks!

**Comment by biemond** [comment]
yeah, at some point it will be modbus inverter homey app with main focus solaredge .  it seems growatt cloud side is buggy or cost them too much money , plus it seems this modbus adapter is cheap and works well on most inverters.  always a drama with registers. at least solaredge has it as a default plus good docs.  and not like enphase with local device you can still do max 5 min or pay 250 dollar a month for near realtime data.

**Comment by biemond** [comment]
done https://homey.app/a/solaredge.modbus/test/

**Comment by CaptainVoni** [comment]
Looks like it’s unchanged.
It seems that setCapabilityValue is an async function, so you might need to add „await“ there.

**Comment by biemond** [comment]
that explains , on code itself it is probably done in ms

**Comment by CaptainVoni** [comment]
Just let me know when you added await; I will test again. I'm pretty sure this will solve it.

**Comment by biemond** [comment]
it was not async func so I used a sleep and it works for me 
https://homey.app/a/solaredge.modbus/test/

**Comment by CaptainVoni** [comment]
I can confirm, it works now reliable. Thanks!

I'm just curious: when you tried to use await, did the function simply fail or was the capability not updated before the trigger?

**Comment by biemond** [comment]
with await, the compiler fails because function itself was not async.  so this was the quick solution.
